rust-symcrypt/src/cipher/mod.rs

```diff
@@ -0,0 +1,119 @@
+//! Friendly rust types for BlockCipherTypes. Currently the only supported BlockCipherType is Aes.
+
+use crate::errors::SymCryptError;
+use crate::symcrypt_init;
+use symcrypt_sys;
+
+use std::marker::PhantomPinned;
+use std::pin::Pin;
+use std::sync::Arc;
+
+// export cipher modes
+pub mod cbc;
+pub mod chacha;
+pub mod gcm;
+
+/// 16
+pub const AES_BLOCK_SIZE: u32 = 16;
+
+/// `BlockCipherType` is an enum that enumerates all possible block ciphers that are supported.
+/// Currently the only supported type is `AesBlock`.
+pub enum BlockCipherType {
+    AesBlock,
+}
+struct AesInnerKey {
+    inner: symcrypt_sys::SYMCRYPT_AES_EXPANDED_KEY,
+    _pinned: PhantomPinned,
+}
+
+impl AesInnerKey {
+    pub(crate) fn new() -> Pin<Box<Self>> {
+        Box::pin(AesInnerKey {
+            inner: symcrypt_sys::SYMCRYPT_AES_EXPANDED_KEY::default(),
+            _pinned: PhantomPinned,
+        })
+    }
+
+    pub(crate) fn get_inner_mut(
+        self: Pin<&mut Self>,
+    ) -> *mut symcrypt_sys::SYMCRYPT_AES_EXPANDED_KEY {
+        unsafe {
+            // SAFETY: Accessing the inner state of the pinned data.
+            &mut self.get_unchecked_mut().inner as *mut _
+        }
+    }
+
+    pub(crate) fn get_inner(&self) -> *const symcrypt_sys::SYMCRYPT_AES_EXPANDED_KEY {
+        &self.inner as *const _
+    }
+}
+
+/// `AesExpandedKey` is a struct that represents an expanded AES key. This struct holds no state and is used to encrypt and decrypt data.
+pub struct AesExpandedKey {
+    // Owned expanded key, this has no state, other calls will take reference to this key.
+    pub(crate) expanded_key: Arc<Pin<Box<AesInnerKey>>>,
```

Comment:
Should we give the user a peek at what the expanded key values are? or the key that they passed in? 

Something like: `get_key`? or `get_expanded_key`?

Reply:
I don't see a need at this time.

Reply:
The format of the expanded key in SymCrypt is an implementation detail which callers should not rely on